packages/dd-trace/src/encode/tags-processors.js

```diff
@@ -25,80 +25,59 @@ const MAX_SERVICE_LENGTH = 100
 // MAX_TYPE_LENGTH the maximum length a span type can have
 const MAX_TYPE_LENGTH = 100
 
-const fromEntries = Object.fromEntries || (entries =>
-  entries.reduce((obj, [k, v]) => Object.assign(obj, { [k]: v }), {}))
-
 function truncateToLength (value, maxLength) {
-  if (!value) {
-    return value
-  }
-  if (value.length > maxLength) {
+  if (value && value.length > maxLength) {
     return `${value.slice(0, maxLength)}...`
   }
   return value
 }
 
+// TODO(bengl) Pretty much everything in this file should happen in
+// `format.js`, so that we're not iterating over all the spans and modifying
+// them yet again.
+
 // normally the agent truncates the resource and parses it in certain scenarios (e.g. SQL Queries)
 function truncateSpan (span, shouldTruncateResourceName = true) {
-  return fromEntries(Object.entries(span).map(([key, value]) => {
-    switch (key) {
-      case 'resource':
-        return ['resource', shouldTruncateResourceName ? truncateToLength(value, MAX_RESOURCE_NAME_LENGTH) : value]
-      case 'meta':
-        return ['meta', fromEntries(Object.entries(value).map(([metaKey, metaValue]) =>
-          [truncateToLength(metaKey, MAX_META_KEY_LENGTH), truncateToLength(metaValue, MAX_META_VALUE_LENGTH)]
-        ))]
-      case 'metrics':
-        return ['metrics', fromEntries(Object.entries(value).map(([metricsKey, metricsValue]) =>
-          [truncateToLength(metricsKey, MAX_METRIC_KEY_LENGTH), truncateToLength(metricsValue, MAX_METRIC_VALUE_LENGTH)]
-        ))]
-      default:
-        return [key, value]
+  if (shouldTruncateResourceName && span.resource && span.resource.length > MAX_RESOURCE_NAME_LENGTH) {
+    span.resource = `${span.resource.slice(0, MAX_RESOURCE_NAME_LENGTH)}...`
+  }
+  for (let metaKey in span.meta) {
+    const val = span.meta[metaKey]
+    if (metaKey.length > MAX_META_KEY_LENGTH) {
+      delete span.meta[metaKey]
+      metaKey = `${metaKey.slice(0, MAX_META_KEY_LENGTH)}...`
+    }
+    span.meta[metaKey] = truncateToLength(val, MAX_META_VALUE_LENGTH)
+  }
+  for (let metricsKey in span.metrics) {
+    const val = span.metrics[metricsKey]
+    if (metricsKey.length > MAX_METRIC_KEY_LENGTH) {
+      delete span.metrics[metricsKey]
+      metricsKey = `${metricsKey.slice(0, MAX_METRIC_KEY_LENGTH)}...`
```

Comment:
Some reason you didn't use `truncateToLength` here too?

```suggestion
      metricsKey = truncateToLength(metricsKey, MAX_METRIC_KEY_LENGTH)
```